_drafts/learn-doobie-for-great-good.md

```diff
@@ -0,0 +1,961 @@
+  findActorById.transact(xa)
+}
+```
+
+Although extracting actors in a `List[String]` seems legit at first sight, it's not safe in a real-world scenario. In fact, the number of extracted rows could be huge to not fit inside the memory allocated to the application. For this reason, we should use a `Stream` instead of a `List`. **Doobie integrates smoothly with the functional streaming library [fs2](https://fs2.io)**. Again, describing how fs2 works is behind the scope of this article, and we just focus on how to use it with Doobie.
+
+For example, let's change the above example to use the streaming API:
+
+```scala
+val actorsNamesStream: fs2.Stream[doobie.ConnectionIO, String] = 
+  sql"select name from actors".query[String].stream
+```
+
+The `stream` method on the type `Query0` returns a `Stream[ConnectionIO, A]` which means a stream containing instances of type `A`, wrapped in an effect of type `ConnectionIO` (for a brief explanation of the Effect pattern, please refer to [The Effect Pattern](https://blog.rockthejvm.com/zio-fibers/#2-the-effect-pattern)).
+
+Once we obtained an instance of a `Stream`, we can decide to return it to the caller as it is, or to compile it into a finite type, such as `List[String]` or `Vector[String]`:
+
+```scala
+val actorsNamesList: IO[List[String]] = actorsNamesStream.compile.toList.transact(xa)
+```
+
+Clearly, there are more than single-column queries. In fact, Doobie can handle multi-column queries as well. For example, let's query the ids and names of all the actors and return them as a tuple:
+
+```scala
+def findAllActorsIdsAndNamesProgram: IO[List[(Int, String)]] = {
+  val query: doobie.Query0[(Int, String)] = sql"select id, name from actors".query[(Int, String)]
+  val findAllActors: doobie.ConnectionIO[List[(Int, String)]] = query.to[List]
+  findAllActors.transact(xa)
+}
+```
+
+We can map the query result inside an instance of an `HList` or inside an `Option`. However, as we can imagine, the most helpful mapping of returned columns is directly into a class. Let's say we want to store the information of extracted actors directly into instances of the class `Actor` class:
+
+```scala
+def findAllActorsProgram: IO[List[Actor]] = {
+  val findAllActors: fs2.Stream[doobie.ConnectionIO, Actor] =
+    sql"select id, name from actors".query[Actor].stream
+  findAllActors.compile.toList.transact(xa)
+}
+```
+
+**Doobie can map the tuple of extracted columns directly into a `case class`**. For now, let's say that the mapping between the extracted tuple and the properties of the case class must be one-to-one. In the last part of the article, we will introduce the type classes that allow the conversion of a tuple into a case class.
+
+The last aspect we left about selecting information from a table is to parameterize the query with parameters. Fortunately, the `sql` interpolator works smoothly with parameters, using the exact mechanism used by Scala native `String` interpolation:
```

Comment:
The last aspect we left about -> One last thing about
parameterize the query with parameters -> add parameters to the query